table/fib-strategy-hashtable.go

```diff
@@ -0,0 +1,410 @@
+/* YaNFD - Yet another NDN Forwarding Daemon
+ *
+ * Copyright (C) 2022 Danning Yu.
+ *
+ * This file is licensed under the terms of the MIT License, as found in LICENSE.md.
+ */
+
+// This file implements a hash table version of the FIB, as outlined in section 2.3 of:
+// W. So, A. Narayanan and D. Oran,
+// "Named data networking on a router: Fast and DoS-resistant forwarding with hash tables,"
+// Architectures for Networking and Communications Systems, 2013, pp. 215-225,
+// doi: 10.1109/ANCS.2013.6665203.
+
+package table
+
+import (
+	"sync"
+
+	"github.com/cespare/xxhash"
+	"github.com/named-data/YaNFD/ndn"
+)
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+
+	return b
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+
+	return b
+}
+
+type virtualDetails struct {
+	// isVirtual marks whether this a virtual node, as defined in the paper.
+	// If there exists a name with length greater than M, then the node is
+	// marked as virtual.
+	isVirtual bool
+
+	// md is the max depth associated with this virtual node, as defined in the paper.
+	md int
+}
+
+// FibStrategy Tree represents a tree implementation of the FIB-Strategy table.
+type FibStrategyHashTable struct {
+	// m is the name length for virtual nodes, as defined in the paper
+	// Must be a positive value
+	// TODO: Allow for the value of m to be changed after the FIB is created.
+	//       This would require deleting and recreating virtual nodes.
```

Comment:
That's because they cannot give a useful `M` value and think that may depend on scenarios. I don't think changing a critical parameter of a data structure dynamically is a good idea: you need to reconstruct the whole structure anyway.